docs/src/nsm-proxy-gui.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.2.
.TH NSM-PROXY-GUI "1" "October 2022" "nsm-proxy-gui Version 1.6.1" "User Commands"
.SH NAME
nsm-proxy-gui \- manual page for nsm-proxy-gui Version 1.6.1
.SH DESCRIPTION
nsm\-proxy\-gui \- GUI for nsm\-proxy, a wrapper for executables without direct NSM\-Support.
.SS "Usage:"
.IP
nsm\-proxy\-gui \fB\-\-help\fR
nsm\-proxy\-gui \fB\-\-connect\-to\fR
.SH OPTIONS
.TP
\fB\-\-help\fR
Show this screen
.TP
\fB\-\-connect\-to\fR
Connect to running nsm\-proxy
.PP
nsmd\-proxy\-gui is usually not called by the user directly,
but autostarted when nsm\-proxy is added to a session (through a GUI).
.SH "REPORTING BUGS"
https://github.com/jackaudio/new-session-manager/issues
.SH COPYRIGHT
up to 2020:
Jonathan Moore Liles https://non.tuxfamily.org/

from 2020:
Nils Hilbricht et al. https://new-session-manager.jackaudio.org
.SH "SEE ALSO"
The  full  documentation for NSM is maintained as html site in your systems doc-dir.
For example:
    xdg-open file:///usr/share/doc/new-session-manager/index.html

The documentation can also be found online https://new-session-manager.jackaudio.org


CHANGELOG
#Changelog
Format: Double ## for a version number followed by a space, ISO-Date, Semantic Versioning:
## YYYY-MM-DD major.minor.patch
Two empty lines before the next entry.
External contributors notice at the end of the line: (LastName, FirstName / nick)


## 2022-10-15 1.6.1
Absolutely make sure that all clients of the session are closed when closing the session.
 This solves the invisible, hidden clients still running even after nsmd quit itself.
 Decrease wait time for such clients at session quit from extreme 60s to very long 30s.
For lockfiles fall back to hardcoded /run/user/<user-id> when $XDG_RUNTIME_DIR is not available on the system.


## 2022-04-15 1.6.0
nsmd:
 Now follows the XDG Base Directory Specifications.
 Default session directory moved from ~/NSM Sessions/ to $XDG_DATA_HOME/nsm/ (see issue #gh-15)
 The old path ~/NSM Sessions/ is still supported and has priority, for now. This may be switched off in the future.

 Lockfiles fixed (see issue #gh-31)
 Lockfiles are now in $XDG_RUNTIME_DIR/nsm/
 Lockfiles now each contain the session path, the osc NSM_URL and the nsmd PID
 One daemon file for each currently running nsmd is created in $XDG_RUNTIME_DIR/nsm/d/ containing the osc url. This enables discovery of running daemons.

 New section in the API documentation for the above.
 Handle write-protected session files and related errors on save. They will not crash the daemon anymore.
 Fixes and guards against trying to load non-existing sessions and creating new sessions under existing names
 Handle various crashes-on-exit and replace them with controlled exits.

Jackpatch Version 1.0.0 (previously 0.2.0):
  Jackpatch will finally not "forget" connections anymore. See #gh-74
  Reduce verbosity level of log ouput.
  Document 'hidden' standalone (no NSM) command line mode in --help
  Handle SIGNALs even when in standalone mode
  Add a jackpatch desktop file with X-NSM-Capable=true and X-NSM-Exec=jackpatch and NoDisplay=true

NSM-Proxy:
    Add a nsm-proxy desktop file with X-NSM-Capable=true and X-NSM-Exec=nsm-proxy and NoDisplay=true


## 2022-01-15 1.5.3
Add [jackpatch] to terminal log output of jackpatch.
Remove hardcoded ANSI colors from terminal log output


## 2021-07-15 1.5.2
pynsm2 library: Fixed a rare crash, where the hostname must be case sensitive but Pythons urlparse forced lower-case.


## 2021-03-19 1.5.1
Web-URLs changed to https://new-session-manager.jackaudio.org and https://github.com/jackaudio/new-session-manager
No codechanges.


## 2021-01-15 1.5.0
WARNING: Next scheduled release (2021-04-15) will switch the default session root
to $XDG_DATA_HOME ( default on most distributions: ~/.local/share/nsm/ )
With the next release prepare to do one of the following:
*  Move old sessions to the new root directory (preferred)
*  Symlink "~/NSM Sessions" to the new root directory
*  use the nsmd --session-root commandline argument.

nsmd:
 Fix session discovery to not report nested sessions anymore. Also more robust file system error handling.
 Command line option --quiet: Suppress messages except warnings and errors
 Protect against orphaned clients or daemons when the server, or even a GUI, crashes.
 Replace cowboy-slang in info-level OSC with descriptive, technical messages.

Legacy-GUI:
 Fix manpage description and usage with the correct executable name
 Fix resizing to very small and back. ( / TheGreatWhiteShark )

NSM-Proxy:
  Multiple layout and style fixes. Better texts for beginners.

API:
 NSM_API_VERSION_PATCH from 0 to 1 (1.1.0 -> 1.1.1)
 Please see API document chapter "Changes in API Version 1.1.1"

Extras:
 This repository now contains extras (libraries, programs, documentation etc.)
 Extras are technically not connected to the main programs of this repository.
 There is no dependency to any "extra" nor any license implications.
 Please read extras/README.md.

 nsm.h was moved to extras/nsm.h
 "extras/pynsm" is now a part of NEW-SM. It was a standalone git repo until now.


## 2020-07-15 1.4.0
Add documentation and manpages.

Legacy-GUI:
 Overhaul look and feel.
 Rewrite labels and buttons with unambiguous descriptions.
 Protect text-input dialog windows from empty strings, like "Add New Client" or "New Session"
 Scale icons, support more icon formats.
 Show all icons and buttons when attaching to a running nsmd session
 Various small fixes.
 Always show correct session name, no matter how the session was loaded or how the GUI was started

nsmd:
 NSM_API_VERSION_MINOR from 0 to 1 (1.0 -> 1.1)
 Repair nsmd to correctly send client data when running headless and a GUI announces later.
 ClientId generation now prevent collision with existing IDs.
 nsmd command line option --load-session to directly load one (Berkelder, Rik)
 Better detection of clients that failed to launch leads to faster session startup (by 5 seconds)
 Users get informed by client-label if an executable is not present on the system or permission denied
 Fixed reply for listing sessions from a plain "Done." to proper reply path with empty string as terminal symbol  "/reply", "/nsm/server/list", ""
 Fix operation reply to last treated client instead to reply to sender (Picot, Mathieu  / houston)
 /nsm/gui/session/name send consistent session name/relative-path pair to the annouced GUI, no matter how the session was loaded.

nsm.h
 :optional-gui: support to nsm.h, for other applications to include and use. (Meyer, Hermann / brummer)


## 2020-06-20 1.3.2
Rename new-session-manager executable to nsm-legacy-gui to prevent future confusion.


## 2020-06-20 1.3.1
Add header copyright even to unchanged files to adhere to stricter packaging requirements.
Meson can now switch off individual executables and will stop/error when trying to build without dependencies.


## 2020-06-17 1.3.0
Rebranding to "new session manager"
Upstream GUI tools "non-session-manager" and "nsm-proxy" converted to standard FLTK instead of a custom toolkit
Changed build system to meson
License upgraded to GPLv3

src/nsm-proxy-gui.cpp

/*******************************************************************************/
/* Copyright (C) 2008-2020 Jonathan Moore Liles (as "Non-Session-Manager")     */
/* Copyright (C) 2020- Nils Hilbricht                                          */
/*                                                                             */
/* This file is part of New-Session-Manager                                    */
/*                                                                             */
/* New-Session-Manager is free software: you can redistribute it and/or modify */
/* it under the terms of the GNU General Public License as published by        */
/* the Free Software Foundation, either version 3 of the License, or           */
/* (at your option) any later version.                                         */
/*                                                                             */
/* New-Session-Manager is distributed in the hope that it will be useful,      */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of              */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the               */
/* GNU General Public License for more details.                                */
/*                                                                             */
/* You should have received a copy of the GNU General Public License           */
/* along with New-Session-Manager. If not, see <https://www.gnu.org/licenses/>.*/
/*******************************************************************************/

#pragma GCC diagnostic ignored "-Wunused-parameter"


#define _MODULE_ "nsm-proxy-gui"

#define APP_NAME "NSM Proxy"
#define APP_TITLE "NSM Proxy"

#include <FL/Fl_File_Chooser.H>
#include <FL/Fl_Text_Display.H>
#include "NSM_Proxy_UI.H"
#include <lo/lo.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>


lo_server losrv;
lo_address nsmp_addr;

static NSM_Proxy_UI *ui;

static char *client_error;

int
osc_update ( const char *path, const char *types, lo_arg **argv, int argc, lo_message msg, void *user_data )
{
    //Updates are arriving one OSC message at a time.

    printf( "Got update for %s\n", path );

    Fl::lock();

    if (!strcmp( path, "/nsm/proxy/label" ))
        ui->label_input->value( &argv[0]->s );
    else if (!strcmp( path, "/nsm/proxy/arguments" ))
        ui->arguments_input->value( &argv[0]->s );
    else if (!strcmp( path, "/nsm/proxy/executable" )) {
        ui->executable_input->value( &argv[0]->s );
        if ( strcmp( &argv[0]->s , "") ) {
            //We want to avoid that the button is always labeled 'Start', creating the
            //false impression that a new sub-client instance is started each time you press it.
            //If the string is empty there is no program running at the moment.
            //This does not cover all cases but gets us there 90%, which is good enough for something cosmetic.
            ui->start_button->label("Ok");
        }
    }
    else if (!strcmp( path, "/nsm/proxy/config_file" ))
        ui->config_file_input->value( &argv[0]->s );
    else if (!strcmp( path, "/nsm/proxy/save_signal" ))
    {
        if ( argv[0]->i == SIGUSR1 )
            ui->save_signal_choice->value( 1 );
        else if ( argv[0]->i == SIGUSR2 )
            ui->save_signal_choice->value( 2 );
        else if ( argv[0]->i == SIGINT )
            ui->save_signal_choice->value( 3 );
        else
            ui->save_signal_choice->value( 0 );
    }
    else if (!strcmp( path, "/nsm/proxy/stop_signal" ))
    {
        if ( argv[0]->i == SIGTERM )
            ui->stop_signal_choice->value( 0 );
        else if ( argv[0]->i == SIGINT )
            ui->stop_signal_choice->value( 1 );
        else if ( argv[0]->i == SIGHUP )
            ui->stop_signal_choice->value( 2 );
    }
    if (!strcmp( path, "/nsm/proxy/client_error" ))
    {
        if ( client_error != NULL )
            free(client_error);

        client_error = NULL;

        if ( strlen(&argv[0]->s) > 0 )
            client_error = strdup(&argv[0]->s);
    }

    Fl::unlock();

    return 0;
}


void
init_osc ( const char *osc_port )
{

    lo_server_thread loth = lo_server_thread_new( osc_port, NULL );
    losrv = lo_server_thread_get_server( loth );

//error_handler );

    char *url = lo_server_get_url(losrv);
    printf("OSC: %s\n",url);
    free(url);

    /* GUI */

    lo_server_thread_add_method( loth, "/nsm/proxy/executable", "s", osc_update, NULL );
    lo_server_thread_add_method( loth, "/nsm/proxy/arguments", "s", osc_update, NULL );
    lo_server_thread_add_method( loth, "/nsm/proxy/config_file", "s", osc_update, NULL );
    lo_server_thread_add_method( loth, "/nsm/proxy/label", "s", osc_update, NULL );
    lo_server_thread_add_method( loth, "/nsm/proxy/save_signal", "i", osc_update, NULL );
    lo_server_thread_add_method( loth, "/nsm/proxy/stop_signal", "i", osc_update, NULL );
    lo_server_thread_add_method( loth, "/nsm/proxy/client_error", "s", osc_update, NULL );

    lo_server_thread_start( loth );
}

/*****************/
/* GUI Callbacks */
/*****************/

void
handle_kill ( Fl_Widget *o, void *v )
{
    lo_send_from( nsmp_addr, losrv, LO_TT_IMMEDIATE, "/nsm/proxy/kill", "" );
}

void
handle_start ( Fl_Widget *o, void *v )
{
    //Executed when Start Button is clicked.
    //"/nsm/proxy/start" for the sub-client is always sent, no matter if the software already runs.
    //nsm-proxy.cpp handles the redundancy.
    lo_send_from( nsmp_addr, losrv,  LO_TT_IMMEDIATE, "/nsm/proxy/start", "sss",
                  ui->executable_input->value(),
                  ui->arguments_input->value(),
                  ui->config_file_input->value() );
}

void
handle_label ( Fl_Widget *o, void *v )
{
    lo_send_from( nsmp_addr, losrv, LO_TT_IMMEDIATE, "/nsm/proxy/label", "s",
                  ui->label_input->value() );
}

void
handle_executable ( Fl_Widget *o, void *v )
{
    //Autofill the label field when the executable name is edited
    ui->label_input->value( ui->executable_input->value() );
    //This does not trigger handle_label so we call it manually
    lo_send_from( nsmp_addr, losrv, LO_TT_IMMEDIATE, "/nsm/proxy/label", "s",
                  ui->label_input->value() );
}


void
handle_config_file ( Fl_Widget *o, void *v )
{
}

void
handle_config_file_browse ( Fl_Widget *o, void *v )
{
    const char * file = fl_file_chooser( "Pick file", "*", NULL, 1 );

    ui->config_file_input->value( file );
}

void
handle_save_signal ( Fl_Widget *o, void *v )
{
    int sig = 0;

    const char* picked = ui->save_signal_choice->mvalue()->label();

    if ( !strcmp( picked, "SIGUSR1" ) )
        sig = SIGUSR1;
    else if ( !strcmp( picked, "SIGUSR2" ) )
        sig = SIGUSR2;
    else if ( !strcmp( picked, "SIGINT" ) )
        sig = SIGINT;

    lo_send_from( nsmp_addr, losrv,  LO_TT_IMMEDIATE,"/nsm/proxy/save_signal", "i",
                  sig );
}

void
handle_stop_signal ( Fl_Widget *o, void *v )
{
    int sig = SIGTERM;

    const char* picked = ui->stop_signal_choice->mvalue()->label();

    if ( !strcmp( picked, "SIGTERM" ) )
        sig = SIGTERM;
    else if ( !strcmp( picked, "SIGINT" ) )
        sig = SIGINT;
    else if ( !strcmp( picked, "SIGHUP" ) )
        sig = SIGHUP;

    lo_send_from( nsmp_addr, losrv,  LO_TT_IMMEDIATE,"/nsm/proxy/stop_signal", "i",
                  sig );
}

void
connect_ui ( void )
{
    ui->executable_input->callback( handle_executable, NULL );
    ui->config_file_input->callback( handle_config_file, NULL );
    ui->kill_button->callback( handle_kill, NULL );
    ui->start_button->callback( handle_start, NULL );
    ui->save_signal_choice->callback( handle_save_signal, NULL );
    ui->stop_signal_choice->callback( handle_stop_signal, NULL );
    ui->label_input->callback( handle_label, NULL );
    ui->config_file_browse_button->callback( handle_config_file_browse, NULL );
}



void cb_dismiss_button ( Fl_Widget *w, void *v )
{
    w->window()->hide();
}

void
check_error ( void *v )
{
    if ( client_error )
    {
        {
            Fl_Double_Window *o = new Fl_Double_Window(600,300+15,"Abnormal Termination");
            //Create a new floating Window that shows an error message.
            {
                Fl_Box *o = new Fl_Box(0+15,0+15,600-30,50);
                o->box(FL_BORDER_BOX);
                o->color(FL_RED);
                o->labelcolor(FL_WHITE);
                o->align(FL_ALIGN_CENTER|FL_ALIGN_WRAP);
                o->copy_label( client_error );
            }
            {
                Fl_Text_Display *o = new Fl_Text_Display(0+15,50+15,600-30,300-75-30);
                o->buffer(new Fl_Text_Buffer());
                o->buffer()->loadfile( "error.log" );
            }
            {
                Fl_Button *o = new Fl_Button(600-75-15,300-25,75,25,"Dismiss");
                o->callback(cb_dismiss_button,0);
            }

            o->show();
        }

        free(client_error);
        client_error = NULL;
    }

    Fl::repeat_timeout( 0.5f, check_error, v );
}

int
main ( int argc, char **argv )
{
    //The NSM-Proxy GUI is a client that communicates with another binary, nsm-proxy via OSC.
    //This GUI executable is actually closed and restarted each time you show/hide the GUI.
    //In other words: it is not a persistent GUI state.


    //Command line parameters
    const char * gui_url = NULL;
    static struct option long_options[] =
    {
        { "connect-to", required_argument, 0, 'u' },
        { "help", no_argument, 0, 'h' },
        { 0, 0, 0, 0 }
    };
    int option_index = 0;
    int c = 0;
    while ( ( c = getopt_long_only( argc, argv, "", long_options, &option_index  ) ) != -1 )
    {
        switch ( c )
        {
            case 'u':
            {
                gui_url = optarg;
                break;
            }

            case 'h':
            {
                const char *usage =
                "nsm-proxy-gui - GUI for nsm-proxy, a wrapper for executables without direct NSM-Support.\n\n"
                "Usage:\n"
                "  nsm-proxy-gui --help\n"
                "  nsm-proxy-gui --connect-to\n"
                "\n"
                "Options:\n"
                "  --help                Show this screen\n"
                "  --connect-to          Connect to running nsm-proxy\n"
                "\n\n"
                "nsmd-proxy-gui is usually not called by the user directly,\n"
                "but autostarted when nsm-proxy is added to a session (through a GUI).\n"
                "";
                puts ( usage );
                exit(0);
                break;
            }
        }
    }

    if ( gui_url == NULL )
        exit(1);

    init_osc( NULL );

    nsmp_addr = lo_address_new_from_url( gui_url );

    if  ( ! nsmp_addr )
        exit(1);

    printf( "Connecting to nsm-proxy at: %s\n", gui_url );

    ui = new NSM_Proxy_UI;

    Fl::scheme( "gtk+" );
    Fl::visual(FL_DOUBLE|FL_INDEX); // FLKT Double_Window: "higly recommended […] put before the first show() of any window in your program"

    Fl_Double_Window *w = ui->make_window();

    connect_ui();

    lo_send_from( nsmp_addr, losrv,  LO_TT_IMMEDIATE, "/nsm/proxy/update", "" );

    //The config file option allows the user to choose a different config file for the proxy settings.
    //This does more harm than good, so we hide the gui field.
    ui->config_file_input->hide();

    w->show();

    Fl::lock();

    //Setting colors only after main window creation.
    //We keep them all in once place instead of setting them in the widgets
    //Colors are the same as nsm-legacy-gui.cpp . If one changes you need to change the other by hand.


    Fl::set_color( 55, 223, 237, 255 ); //Override FLUID palette with RGB Value. 55 is label text. Same as FL_FOREGROUND_COLOR
    Fl::set_color( 41, 55, 61, 69 ); //Override FLUID palette with RGB Value. 41 is label background

    Fl::set_color( FL_DARK1, 37, 40, 45 ); //Main window background
    Fl::set_color( FL_BACKGROUND_COLOR, 37, 40, 45 ); //These are the colors used as backgrounds by almost all widgets and used to draw the edges of all the boxtypes.
    Fl::set_color( FL_BACKGROUND2_COLOR, 55, 61, 69 ); //This color is used as a background by Fl_Input and other text widgets.
    Fl::set_color( FL_FOREGROUND_COLOR, 223, 237, 255 );
    Fl::set_color( FL_INACTIVE_COLOR, 255, 0, 0 ); // Not used
    Fl::set_color( FL_SELECTION_COLOR, 80, 84, 92 ); // e.g. the currently selected session
    Fl::reload_scheme();


    Fl::add_timeout( 0.5f, check_error, NULL );

    Fl::run();

    return 0;
}
